[android] Remove SpongyCastle

sjchmiela opened the pull request:
# Why

May fix https://github.com/expo/expo/issues/5688.

# How

From what I understood from the code (SpongyCastle integration was in the Android codebase from day one, so no PR description for me):
- we verify whether the app that is being served actually is served by an authenticated Expo user (`expo-cli` sends an authenticated request to Expo servers to sign the manifest under given username with a private key)
- X.509 is used as the format of the key
- since not every Android platform provides a `KeyFactory` for `X.509` keys, SpongyCastle was added to ensure X.509 factory works on every device

What I found:
- no online resource uses `X.509` as an argument to `KeyFactory.getInstance`.
- `X.509` was supported only in Android APIs 1-8 ([source](https://developer.android.com/guide/topics/security/cryptography#SupportedKeyFactory))
- `RSA` is supported in Android API 1+ ([source](https://developer.android.com/guide/topics/security/cryptography#SupportedKeyFactory))
- if we substitute `X.509` with `RSA` in the `KeyFactory.getInstance`, the signature is still validated
- if we substitute `X.509` with `DSA`, the signature validation fails
- after these changes our code looks just like `verify` method from `godlibrary`: https://github.com/abook23/godlibrary/blob/master/src/main/java/com/god/util/RSAUtlis.java#L227-L243

# Test Plan

Running experiences works. `SecureStore` NCL screen still works as expected.

Values saved in `SecureStore` are persisted between `master` and this branch.

sjchmiela committed:
[android] Remove SpongyCastle

sjchmiela committed:
[android] Change algorithm used for generating key for manifest verification

sjchmiela committed:
[android] Remove SecureModuleBinding

It was required only for SpongyCastle integration

mczernek reviewed (approved):
🥇

ide commented:
Need to test that SecureStore works on older Android phones (down to Android 5) as well. The simulator can be used to test.

sjchmiela committed:
[expo-secure-store] Pass GCMParameterSpec to AESEncrypter

Old Androids (tested against API 21) try to create an instance of `GCMParameterSpec` using invalid constructor `(byte[], Integer)`, while only `(int, byte[])` is available, resulting in `java.security.spec.InvalidParameterSpecException: no constructor found!`

See https://github.com/bcgit/bc-java/blob/eec9382c6ae2b4b3b46f0f419b34bd77be349ec4/prov/src/main/java/org/bouncycastle/jcajce/provider/symmetric/AES.java#L435 (issue has been fixed in https://github.com/bcgit/bc-java/commit/507c3917c0c469d10b9f033ad641c1da195e2039#diff-c90a59e823805b6c0dcfeaf7bae65f53)

To work around a bug in BouncyCastle, since we're the ones that create `GCMParameterSpec`, we can reuse it instead of asking the Cipher to create it again. This mitigates the problem.

sjchmiela committed:
[expo-secure-store] Fix all lint problems

sjchmiela commented:
I had to pass the instance of `GCMParameterSpec` to `AESEncypter` not to bump into a BouncyCastle bug (see explanation below), but now I can happily confirm it works on Android emulators: 21, 22, 23, 24, 29. 🎉 

> Old Androids (tested against API 21) try to create an instance of `GCMParameterSpec` using invalid constructor `(byte[], Integer)`, while only `(int, byte[])` is available, resulting in `java.security.spec.InvalidParameterSpecException: no constructor found!` (also reported in https://github.com/expo/expo/issues/5688)
> 
> See [bcgit/bc-java/…/org/bouncycastle/jcajce/provider/symmetric/AES.java#L435](https://github.com/bcgit/bc-java/blob/eec9382c6ae2b4b3b46f0f419b34bd77be349ec4/prov/src/main/java/org/bouncycastle/jcajce/provider/symmetric/AES.java#L435) (issue has been fixed in [bcgit/bc-java@507c391](https://github.com/bcgit/bc-java/commit/507c3917c0c469d10b9f033ad641c1da195e2039#diff-c90a59e823805b6c0dcfeaf7bae65f53))
> 
> To work around a bug in BouncyCastle, since we're the ones that create `GCMParameterSpec`, we can reuse it instead of asking the Cipher to create it again. This mitigates the problem.
>
> Source: https://github.com/expo/expo/pull/6464/commits/5506e3709890c34f98e73fe0fb95c4b93e042164

ide commented on packages/expo-secure-store/android/src/main/java/expo/modules/securestore/SecureStoreModule.java:
```diff
@@ -128,7 +128,7 @@ private void setItemImpl(String key, String value, ReadableArguments options, Pr
 
     String encryptedItemString = encryptedItem.toString();
-    if (encryptedItemString == null) {
+    if ("".equals(encryptedItemString)) {
```
> Make a JSON text of this JSONObject. For compactness, no whitespace is added. If this would not result in a syntactically correct JSON text, then null will be returned instead.

https://stleary.github.io/JSON-java/org/json/JSONObject.html#toString-int-

sjchmiela replied:
Whoops, sorry, I was so happy it finally worked that I decided to fix all lint warnings, some of them too mechanically…

ide commented on packages/expo-secure-store/android/src/main/java/expo/modules/securestore/SecureStoreModule.java:
```diff
@@ -466,9 +470,9 @@ public String decryptItem(JSONObject encryptedItem, KeyStore.SecretKeyEntry secr
     protected Context mContext;
     private AESEncrypter mAESEncrypter;
-    protected SecureRandom mSecureRandom;
+    /* package */ SecureRandom mSecureRandom;
```
If something outside this class needs to access it, it should call `getSecureRandom()`

sjchmiela replied:
Changed to `private`, nothing outside of this class accesses it.

ide commented on packages/expo-secure-store/android/src/main/java/expo/modules/securestore/SecureStoreModule.java:
```diff
@@ -396,18 +400,18 @@ public JSONObject createEncryptedItem(String plaintextValue, KeyStore keyStore,
       byte[] ciphertextBytes = cipher.doFinal(plaintextBytes);
       String ciphertext = Base64.encodeToString(ciphertextBytes, Base64.DEFAULT);
 
-      GCMParameterSpec gcmSpec = cipher.getParameters().getParameterSpec(GCMParameterSpec.class);
```
This ensures we use the same spec that the cipher uses -- it would be better to keep this if possible, and if we remove this I think we need to add an assertion that the cipher's algorithm matches the spec object that is passed in.

sjchmiela replied:
On Android 23+ it works exactly as it did (we use `getParameterSpec()`, see https://github.com/expo/expo/pull/6464/files#diff-c4fb6019b00e115649c6aafbcb5d3d82R405).

The change had to be made for Androids below 23, which seem to ship with a faulty BouncyCastle version which tries to instantiate `GCMParameterSpec` with an invalid constructor (see [fixing commit](https://github.com/bcgit/bc-java/commit/507c3917c0c469d10b9f033ad641c1da195e2039#diff-c90a59e823805b6c0dcfeaf7bae65f53)).

I think `Cipher` would [throw an exception](https://android.googlesource.com/platform/libcore/+/d416195/luni/src/main/java/javax/crypto/Cipher.java#649) if it couldn't fulfill the `GCMParameterSpec` we provide him in `aesCipher.init(Cipher.ENCRYPT_MODE, secretKey, gcmSpec);`, so I don't believe that the assertion will ever fail, but I have created such sanity-checking path in code. 🙂

sjchmiela replied:
When you approve of these changes I will backport the changes to older SDKs. (reminder for myself)

ide commented on packages/expo-secure-store/android/src/main/java/expo/modules/securestore/SecureStoreModule.java:
```diff
@@ -494,7 +497,7 @@ public String getKeyStoreAlias(ReadableArguments options) {
           .setEndDate(new Date(Long.MAX_VALUE))
           .build();
 
-      KeyPairGenerator keyPairGenerator = KeyPairGenerator.getInstance(KeyProperties.KEY_ALGORITHM_RSA, keyStore.getProvider());
```
I prefer to use the constants -- is this causing an Android API compatibility issue?

sjchmiela replied:
I prefer to use the constants too, however it triggered the `InlinedApi` lint warning:

> This check scans through all the Android API field references in the application and flags certain constants, such as static final integers and Strings, which were introduced in later versions. *These will actually be copied into the class files rather than being referenced, which means that the value is available even when running on older devices.* In some cases that's fine, and in other cases it can result in a runtime crash or incorrect behavior. It depends on the context, so consider the code carefully and decide whether it's safe and can be suppressed or whether the code needs to be guarded.

Basing on the bolded text I have suppressed the warnings on the offending lines.

sjchmiela committed:
[expo-secure-store] Sanity check for Cipher parameters

ide reviewed (approved):
Thanks, looks nice.

sjchmiela committed:
[android] Backport changes to older SDKs

sjchmiela committed:
[changelog] Add changelog entry

sjchmiela committed:
Merge branch 'master' into @sjchmiela/remove_spongycastle
